packages/transaction-controller/src/utils/batch.ts

```diff
@@ -0,0 +1,157 @@
+import {
+  TransactionEnvelopeType,
+  type TransactionBatchRequest,
+  type TransactionBatchResult,
+  type TransactionParams,
+  TransactionType,
+} from '../types';
+
+type AddTransactionBatchRequest = {
+  addTransaction: TransactionController['addTransaction'];
+  getChainId: (networkClientId: string) => Hex;
+  getEthQuery: (networkClientId: string) => EthQuery;
+  messenger: TransactionControllerMessenger;
+  request: TransactionBatchRequest;
+};
+
+type IsAtomicBatchSupportedRequest = {
+  address: Hex;
+  getEthQuery: (chainId: Hex) => EthQuery;
+  messenger: TransactionControllerMessenger;
+};
+
+const log = createModuleLogger(projectLogger, 'batch');
+
+/**
+ * Add a batch transaction.
+ *
+ * @param request - The request object including the user request and necessary callbacks.
+ * @returns The batch result object including the batch ID.
+ */
+export async function addTransactionBatch(
+  request: AddTransactionBatchRequest,
+): Promise<TransactionBatchResult> {
+  const {
+    addTransaction,
+    getChainId,
+    messenger,
+    request: userRequest,
+  } = request;
+
+  const { from, networkClientId, requireApproval, transactions } = userRequest;
+
+  log('Adding', userRequest);
+
+  const chainId = getChainId(networkClientId);
+  const ethQuery = request.getEthQuery(networkClientId);
+  const isChainSupported = doesChainSupportEIP7702(chainId, messenger);
+
+  if (!isChainSupported) {
+    log('Chain does not support EIP-7702', chainId);
+    throw rpcErrors.internal('Chain does not support EIP-7702');
+  }
+
+  const { delegationAddress, isSupported } = await isAccountUpgradedToEIP7702(
+    from,
+    chainId,
+    messenger,
+    ethQuery,
+  );
+
+  log('Account', { delegationAddress, isSupported });
+
+  if (!isSupported && delegationAddress) {
+    log('Account upgraded to unsupported contract', from, delegationAddress);
+    throw rpcErrors.internal('Account upgraded to unsupported contract');
+  }
+
+  const nestedTransactions = transactions.map((tx) => tx.params);
+  const batchParams = generateEIP7702BatchTransaction(from, nestedTransactions);
+
+  const txParams: TransactionParams = {
+    from,
+    ...batchParams,
+  };
+
+  if (!isSupported) {
+    const upgradeContractAddress = getEIP7702UpgradeContractAddress(
+      chainId,
+      messenger,
+    );
+
+    if (!upgradeContractAddress) {
+      throw rpcErrors.internal('Upgrade contract address not found');
+    }
+
+    txParams.type = TransactionEnvelopeType.setCode;
+    txParams.authorizationList = [{ address: upgradeContractAddress }];
+  }
+
+  log('Adding batch transaction', txParams, networkClientId);
+
+  const { transactionMeta, result } = await addTransaction(txParams, {
+    nestedTransactions,
+    networkClientId,
+    requireApproval,
+    type: TransactionType.batch,
+  });
+
+  const batchId = transactionMeta.id;
+
+  // Wait for the transaction to be published.
+  await result;
+
+  return {
+    batchId,
+  };
+}
```

Comment:
This does not look like utility function, may be it should have been part of controller itself. Also, does not look a good practice to pass functions, messenger to utility function like above.

Reply:
The main `TransactionController` file is 4000 lines long so this is a fundamental part of our controller architecture to modularise functionality in order to keep it readable, maintainable, and testable.

The practice of passing callbacks and the messenger itself is well established as it's the same as what we do with the controllers themselves in the client, it provides a clean abstraction to provide logic in a decoupled way without duplication or unnecessarily coupling distinct pieces of logic.

Reply:
May be we spend some time to refactor `TransactionController` so the main methods can still live in controller and we move helper code to utils.